torch/_dynamo/variables/misc.py

```diff
@@ -767,16 +767,20 @@ def func(tx, item, acc):
             return variables.ListIteratorVariable(
                 items, mutable_local=MutableLocal(), **options
             )
-        elif (
-            self.value is functools.wraps
-            and not kwargs
-            and len(args) == 1
-            and args[0].source
-        ):
+        elif self.value is functools.wraps and not kwargs and len(args) == 1:
 
             def wraps(fn):
+                # args[0] is the function to wrap.
+                # fn is the function being wrapped
+                # Eg.
+                #   wraps(foo)
+                #   def bar():
+                #      pass
+                # Here : args[0] -> foo, fn -> bar
                 if isinstance(fn, variables.NestedUserFunctionVariable):
-                    return fn.clone(wraps_source=args[0].source)
+                    # `wraps_source` should either have a source
+                    # or implement reconstruct.
+                    return fn.clone(wraps_source=args[0])
```

Comment:
Lol, this works but let's not do this. You are taking advantage of the fact that wraps_source is only used in reconstruct. That works until someone tries to access this, thinking its a source, but then its a function! This is how frameworks degrade, imo.

Instead:

1) add a `can_reconstruct` util to base VT 
2) rename `wraps_source` to `reconstructable` and type it to a union of Source and VT
3) maintain the invariant that `codegen` on the NestedUserFunctionVariable w/ a `reconstructable` can reconstruct

Reply:
> (wraps_source is only used for functools.wraps, and the input is always a NestedUserFunctionVariable?)

`wraps_source` is used only for functools.wraps is correct. However input may or may not be a `NestedUserFunctionVariable` (it can also be `UserFunctionVariable` when the wrapped function is in global namespace).

`NestedUserFunctionVariable` itself implements reconstruct (which codegen uses), 
https://github.com/pytorch/pytorch/blob/bc3f0d341aade5f6e444b39fc3c6ee7a8e49d2cf/torch/_dynamo/codegen.py#L147-L155

while in case of `UserFunctionVariable`, it has source and codegen knows how to deal with that.
https://github.com/pytorch/pytorch/blob/bc3f0d341aade5f6e444b39fc3c6ee7a8e49d2cf/torch/_dynamo/codegen.py#L97-L102



Ref:
Usage of `wraps_source`
https://github.com/pytorch/pytorch/blob/bc3f0d341aade5f6e444b39fc3c6ee7a8e49d2cf/torch/_dynamo/variables/functions.py#L524

NOTE: codegen doesn't seem to be typed and deals with a variety of types
https://github.com/pytorch/pytorch/blob/bc3f0d341aade5f6e444b39fc3c6ee7a8e49d2cf/torch/_dynamo/codegen.py#L70-L157

---------

I think renaming `wraps_source` to `wraps_func` makes sense, as it is the function being wrapped. We can type it as `BaseUserFunctionVariable` or  if we want to play safer as `Union[UserFunctionVariable, NestedUserFunctionVariable]`.

Wdyt?

Thanks!

Reply:
On second thought, people can call functools.wraps on anything, right? (it does not need to be a UserFunctionVariable/NestedUserFunctionVariable). If that's the case, then we would want to apply Voz's suggestion of having a `wraps_fun_reconstructible` that is Union[Source, UserFunctionVariable, NestedUserFunctionVariable]

Reply:
@voznesenskym 

Regarding the `can_reconstruct`, I don't think we can easily figure that for a subclass of VariableTracker.

We can't implement it like `can_make_guard` as `reconstruct` expects a codegen object and also codegen mutates the translator when called.
https://github.com/pytorch/pytorch/blob/7f7f6267e99f8dd5990e43dd2301ac37bb66e299/torch/_dynamo/variables/base.py#L241-L246
https://github.com/pytorch/pytorch/blob/7f7f6267e99f8dd5990e43dd2301ac37bb66e299/torch/_dynamo/variables/base.py#L284-L285

Another concern I have is about the notion of reconstructable VT. Codegen doesn't always call `reconstruct` on VariableTracker but handles them case by case (without calling `reconstruct`). So even if we implement `can_reconstruct` like above, it wouldn't still cover all the tracker subclasses that codegen can deal with.
https://github.com/pytorch/pytorch/blob/7f7f6267e99f8dd5990e43dd2301ac37bb66e299/torch/_dynamo/codegen.py#L97-L137

So, with above in mind, one approach could be just adding a flag in VT and then all subclasses have to set it manually. (If we opt for this, I think it should be done in separate PR)

For this PR would it be ok to rename `wraps_source` to `wraps_reconstructible : Union[Source, NestedUserFunctionVariable]` and graph break for other cases (and follow-up separately on reconstruct)?

What do you think about this? Is there anything missing?

Thanks!

Reply:
Gentle ping @voznesenskym :)

Reply:
Gentle ping @voznesenskym